Add HF dataset example pipeline

NielsRogge opened the pull request:
This PR adds a pipeline example that consists of a single component.

The component loads a dataset from the 🤗 hub and uploads it to GCS, leveraging Express. As can be seen in the requirements.txt file of the component, it has a dependency on Express.

PhilippeMoussalli commented on examples/pipelines/hf_dataset_pipeline/README.md:
```diff
@@ -0,0 +1,63 @@
+
+# 1. Implement the components
+
+1. Create a "components" folder where all KubeFlow components will live.
```
I think a lot of this documentation is repeated from something i'v already included before and seems to be more general to kubeflow rather than this specific pipeline.

Could we:

* Define this somewhere else? Maybe on the root level of the pipelines.
* Combine it with the current existing documentation?https://github.com/ml6team/express/blob/main/docs/kubeflow_101.md

PhilippeMoussalli commented on examples/pipelines/hf_dataset_pipeline/README.md:
```diff
@@ -0,0 +1,63 @@
+In this guide, we explain how to implement a KubeFlow pipeline and run it on KubeFlow on GKE (Google Kubernetes Engine).
```
The name 'hf_dataset_pipeline` is a bit abstract and can indicate many things. Can we change it to Stable diffusion data collection or finetuning?

RobbeSneyders replied:
Depends on what we want to showcase. If this example is meant to showcase how to work with a HuggingFace dataset, I think the current name is appropriate.

NielsRogge replied:
Going to keep the name for now

PhilippeMoussalli commented on examples/pipelines/hf_dataset_pipeline/components/load_from_hub/build_image.sh:
```diff
@@ -0,0 +1,28 @@
+# Create full name of the image
+FULL_IMAGE_NAME=${ARTIFACT_PATH}/${IMAGE_NAME}:${IMAGE_TAG}
+echo $FULL_IMAGE_NAME
+
```
I think you can include here `gcloud config set builds/use_kaniko True` to enable it automatically and avoid having the user look through the docs for figuring out this step.
 Can you please include it in all the build_image.sh that exist? you can also update the docs then to just mention that cashing is used

RobbeSneyders replied:
Would propose not to do this in this PR and move to `docker` instead of `gcloud` anyway.

NielsRogge replied:
Will keep this for now, and in a follow-up PR we can update all components to leverage `docker` (in a single script that builds the images for all components)

PhilippeMoussalli commented on examples/pipelines/hf_dataset_pipeline/components/load_from_hub/Dockerfile:
```diff
@@ -0,0 +1,31 @@
+# Copy over src-files of the component
+COPY src /src
+
+# TODO: ADD EXTRA FILES HERE IF NEEDED!
```
This line can be removed

PhilippeMoussalli commented on examples/pipelines/hf_dataset_pipeline/components/load_from_hub/component.yaml:
```diff
@@ -0,0 +1,24 @@
+        image: europe-west4-docker.pkg.dev/soy-audio-379412/soy-audio-379412-default-repository/kubeflow-components/components/load_from_hub:latest
+        command: [
+            python3, main.py,
+            --extra-args,       {inputValue: extra_args},
```
Nitpick alert: could you align the values starting with `{`

PhilippeMoussalli commented on examples/pipelines/hf_dataset_pipeline/components/load_from_hub/requirements.txt:
```diff
@@ -0,0 +1,2 @@
+# Add component requirements here
```
The line can be removed

PhilippeMoussalli commented on examples/pipelines/hf_dataset_pipeline/components/load_from_hub/requirements.txt:
```diff
@@ -0,0 +1,2 @@
+# Add component requirements here
+git+https://github.com/ml6team/express.git@main
```
We might have to think about a more efficient way of using this (e.g. if you make local changes in Express and you want to test them). Could you add a TODO or a ticket for it?

NielsRogge replied:
Yes so @PhilippeMoussalli the examples should always rely on the main branch (not on a certain commit.

I guess we can update the requirements.txt to a certain commit only when creating a new example (for debug purposes), but once it's ready we should update to just the main branch

PhilippeMoussalli commented on examples/pipelines/hf_dataset_pipeline/helpers/lightweight.py:
```diff
@@ -0,0 +1,11 @@
+"""Collection of lightweight components"""
```
This file is not needed I guess

NielsRogge replied:
Will remove

PhilippeMoussalli commented on examples/pipelines/hf_dataset_pipeline/components/load_from_hub/src/main.py:
```diff
@@ -0,0 +1,49 @@
+        caption_dataset = concatenate_datasets([index, caption_dataset])
+        
+        # 2.2) Create data_source dictionary
+        data_sources = {"captions": caption_dataset}
```
ok for the initial loading but we might want to think of a more efficient way of doing this. Right now we're materializing the dataset in both the `data_sources` dict and also the `caption_dataset` which can lead to OoO issues with large datasets

RobbeSneyders replied:
I don't think this is the case since both are referring to the same object.

But the current setup is indeed limited by the local memory, so we should find a way to stream this data instead.

NielsRogge replied:
Does Python duplicate the memory when using a variable in a dict?

PhilippeMoussalli replied:
@RobbeSneyders you're right I think it just references it. I would still opt to use streaming instead 
https://huggingface.co/docs/datasets/stream

NielsRogge replied:
Streaming is mostly useful when you'd like to transform the dataset, not when you want to upload it to the cloud like here, so will keep it like it is. However for our transform components we can leverage streaming

PhilippeMoussalli commented on examples/pipelines/hf_dataset_pipeline/hf_dataset_pipeline.py:
```diff
@@ -0,0 +1,44 @@
+load_from_hub_op = comp.load_component(
+    '/Users/nielsrogge/Documents/python_projects/express/examples/pipelines/hf_dataset_pipeline/components/load_from_hub/component.yaml')
+
+run_id = '{{pod.name}}'
```
This won't work since it's not materialized, Could you add a TODO or a ticket for this and replace with a default value?

RobbeSneyders replied:
What do you mean with this? How should this value be substituted?

PhilippeMoussalli replied:
It should be defined inside the pipeline function for it to take place otherwise you get an error. Haven't dug deep into it yet. Can also use `kfp.dsl.RUN_ID_PLACEHOLDER` somewhere inside Express 

![Screenshot from 2023-03-15 16-59-47](https://user-images.githubusercontent.com/47530815/225371917-3bff4079-f06b-4d70-a18e-dfd40b458f8f.png)

NielsRogge replied:
Ok great!

PhilippeMoussalli reviewed (changes requested):
Reviewed! Overall looks good but left a few comments for a few things to adjust

RobbeSneyders commented on examples/pipelines/hf_dataset_pipeline/components/load_from_hub/README.md:
```diff
@@ -0,0 +1,14 @@
+python3 main.py --extra-args '{"dataset_name": "lambdalabs/pokemon-blip-captions"}' \
+ --output-manifest <local_path> \
+  --metadata-args '{"run_id":"test","component_name":"test_component", \
+  "artifact_bucket":"soy-audio-379412_kfp-artifacts"}'
```
Let's replace the artifact bucket with a placeholder as well.

RobbeSneyders commented on examples/pipelines/hf_dataset_pipeline/components/load_from_hub/README.md:
```diff
@@ -0,0 +1,14 @@
+This component can be run as follows:
```
It documents the arguments specific to this component, so for me this is fine to have here.

PhilippeMoussalli replied:
I think it's more general to data loading in general which will be similar for all dataset frameworks. IMO it should be only mentioned once in a reference doc. I think it would be nice to have a description of the component as well as a schema description for the args in `--extra-args`

RobbeSneyders commented on examples/pipelines/hf_dataset_pipeline/helpers/upload.py:
```diff
@@ -0,0 +1,83 @@
+"""
```
See https://github.com/ml6team/express/pull/15

NielsRogge replied:
Thanks, updated

NielsRogge commented on examples/pipelines/hf_dataset_pipeline/hf_dataset_pipeline.py:
```diff
@@ -0,0 +1,44 @@
+from helpers.upload import compile_and_upload_pipeline
+
+# Load Component
+load_from_hub_op = comp.load_component(
```
Ok, I copied this from the example component

NielsRogge commented:
@PhilippeMoussalli @RobbeSneyders PR is ready for a final review. Addressed all comments + added a transform component to the pipeline which leverages `HFDatasetsTransformComponent`.

PhilippeMoussalli commented on examples/pipelines/hf_dataset_pipeline/components/add_captions/component.yaml:
```diff
@@ -0,0 +1,29 @@
+        image: europe-west4-docker.pkg.dev/soy-audio-379412/soy-audio-379412-default-repository/kubeflow-components/components/add_captions:latest
+        command: [
+            python3, main.py,
+            --input-manifest,       {inputPath: input_manifest},
```
nitpick: can you indent the value and path arguments so that they allign?

PhilippeMoussalli commented on examples/pipelines/hf_dataset_pipeline/components/load_from_hub/src/main.py:
```diff
@@ -0,0 +1,47 @@
+        Returns:
+            HFDatasetsDatasetDraft: a dataset draft that creates a plan for an output datasets/manifest
+        """
+        configure_logging()
```
Include some logging statements in this example as well

PhilippeMoussalli reviewed (approved):
LGTM! Left a few more small comments and then it's ready to ship

NielsRogge commented:
@RobbeSneyders feel free to approve
